src/cache.c

```diff
@@ -0,0 +1,369 @@
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include "cache.h"
+#include <assert.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define MIN(a, b) ((a < b) ? a : b)
+#define GOLDEN_RATIO_32 0x61C88647
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) % (cache_size)
+
+int cache_size, cache_size_bits;
+
+typedef enum {
+    LRU_list,
+    LFU_list,
+    LRU_ghost_list,
+    LFU_ghost_list,
+    list_quantity
+} list_type_t;
+
+struct list_head {
+    struct list_head *prev;
+    struct list_head *next;
+};
+
+struct hlist_head {
+    struct hlist_node *first;
+};
+
+struct hlist_node {
+    struct hlist_node *next, **pprev;
+};
+
+typedef struct arc_entry {
+    void *value;
+    uint32_t key;
+    list_type_t arc_type;
+    struct list_head list;
+    struct hlist_node ht_list;
+} arc_entry_t;
+
+typedef struct hashtable {
+    struct hlist_head *ht_list_head;
+} hashtable_t;
+
+typedef struct cache {
+    struct list_head *list_table[list_quantity];
+    uint32_t list_size[list_quantity];
+    struct hashtable *map;
+    uint32_t capacity;
+    uint32_t lru_capacity;
+} cache_t;
+
+#if defined(__GNUC__)
+#define __LIST_HAVE_TYPEOF 1
+#endif
+
+static inline void INIT_LIST_HEAD(struct list_head *head)
+{
+    head->next = head;
+    head->prev = head;
+}
+
+static inline void list_add(struct list_head *node, struct list_head *head)
+{
+    struct list_head *next = head->next;
+
+    next->prev = node;
+    node->next = next;
+    node->prev = head;
+    head->next = node;
+}
+
+static inline void list_del(struct list_head *node)
+{
+    struct list_head *next = node->next;
+    struct list_head *prev = node->prev;
+
+    next->prev = prev;
+    prev->next = next;
+}
+
+static inline void list_del_init(struct list_head *node)
+{
+    list_del(node);
+    INIT_LIST_HEAD(node);
+}
+
+static inline void list_move(struct list_head *node, struct list_head *head)
+{
+    list_del(node);
+    list_add(node, head);
+}
+
+#define list_entry(node, type, member) container_of(node, type, member)
+
+#define list_last_entry(head, type, member) \
+    list_entry((head)->prev, type, member)
+
+#ifdef __LIST_HAVE_TYPEOF
+#define list_for_each_entry(entry, head, member)                       \
+    for (entry = list_entry((head)->next, __typeof__(*entry), member); \
+         &entry->member != (head);                                     \
+         entry = list_entry(entry->member.next, __typeof__(*entry), member))
+#endif
+
+#define list_for_each_entry_safe(entry, safe, head, member)                \
+    for (entry = list_entry((head)->next, __typeof__(*entry), member),     \
+        safe = list_entry(entry->member.next, __typeof__(*entry), member); \
+         &entry->member != (head); entry = safe,                           \
+        safe = list_entry(safe->member.next, __typeof__(*entry), member))
+
+#undef __LIST_HAVE_TYPEOF
```

Comment:
Don't touch `__LIST_HAVE_TYPEOF`. Ensure that List API works for non-gcc/clang compilers.